src/server/middleware/immutable/pushmf.go

```diff
@@ -59,51 +56,21 @@ func handlePush(req *http.Request) error {
 		return nil
 	}
 
-	// match repository ...
-	total, repos, err := repository.Mgr.List(req.Context(), &q.Query{
-		Keywords: map[string]interface{}{
-			"Name": mf.Repository,
-		},
-	})
-	if err != nil {
-		return err
-	}
-	if total == 0 {
-		return nil
-	}
-
 	// match artifacts ...
-	total, afs, err := artifact.Mgr.List(req.Context(), &q.Query{
-		Keywords: map[string]interface{}{
-			"ProjectID":    mf.ProjectID,
-			"RepositoryID": repos[0].RepositoryID,
-		},
+	af, err := artifact.Ctl.GetByReference(req.Context(), mf.Repository, mf.Tag, &artifact.Option{
+		WithTag: true,
 	})
 	if err != nil {
 		return err
 	}
-	if total == 0 {
+	if af == nil {
 		return nil
 	}
 
-	// match tags ...
-	for _, af := range afs {
-		total, tags, err := tag.Mgr.List(req.Context(), &q.Query{
-			Keywords: map[string]interface{}{
-				"ArtifactID": af.ID,
-			},
-		})
-		if err != nil {
-			return err
-		}
-		if total == 0 {
-			continue
-		}
-		for _, tag := range tags {
-			// push a existing immutable tag, reject the request
-			if tag.Name == mf.Tag {
-				return NewErrImmutable(repoName, mf.Tag)
-			}
+	for _, tag := range af.Tags {
+		// push a existing immutable tag, reject the request
+		if tag.Name == mf.Tag {
```

Comment:
Same here. We can get the artifact with immutable set as true first and check whether the pushing tag already exists and is immutable. 
This can reduce the duplicate code